enhancements/installer/ipi-publishing-status.md

```diff
@@ -0,0 +1,198 @@
+
+### Goals
+
+Expose the cluster publishing status used at install time in the Infrastructure
+object.
+
+### Non-Goals
+
+This enhancement does not cover support for updating the cluster publishing
+status on running clusters. This property is meant to reflect the `publish`
+parameter as it was set in the install-config.yaml.
+
+## Proposal
+
+The Infrastructure API is extended, adding the `Publish` field to the
+`InfrastructureStatus`.
+
+```go
+type InfrastructureStatus struct {
+	// ...
+
+	// publish indicates whether the cluster was installed as public or private.
+	// The default is 'External', which means the cluster will be publicly
+    // available on the internet.
+	// The 'Internal' mode indicates a private cluster, and operators should
+    // configure operands accordingly.
+    // +kubebuilder:default=External
+    // +kubebuilder:validation:Enum=Internal;External
+	Publish PublishMode `json:"publish"`
+}
+```
+
+To reflect the install-config publishing options, `PublishMode` will only
+support to valid options.
+
+```
+// PublishMode defines the publish mode of the cluster at install time.
+type PublishMode string
+
+const (
+	// "Internal" is for operators to configure their asserts as private
+    // when appropriate.
+	InternalPublishMode PublishMode = "Internal"
+
+	// "External" indicates that the cluster is publicly available on the internet.
+	ExternalPublishMode PublishMode = "External"
+)
+```
+
+### Workflow Description
+
+1. The cluster creator install a cluster via IPI, setting `publish: Internal` in installer-config.yaml
+2. The installer bootstraps the cluster, and sets `Publish: Internal` in the Infrastructure status
+3. Operators use this value to configure operands accordingly.
+
+### API Extensions
+
+### Implementation Details/Notes/Constraints [optional]
+
+What are the caveats to the implementation? What are some important details that
+didn't come across above. Go in to as much detail as necessary here. This might
+be a good place to talk about core concepts and how they relate.
+
+The `Publish` field in the infrastructure status will only reflect the initial
+status of the cluster. Individual components (like ingress) may be changed on
+day-2 to for example turn a public cluster private, and the `Publish` field in
+the Infrastructure status will remain untouched.
+
+#### Hypershift [optional]
+
+Hypershift publishing options are not an exact match for the ones provided by
+the installer in installer-config.yaml.
+
+Hypershift options (limited to AWS) are:
+
+* `Public`
+* `PublicAndPrivate`
+* `Private`
+
+`PublicAndPrivate` does not translate well to OpenShift. It allows public API
+server access and private node communication with the control plane, making
+the cluster appear public to end-users, while securing control plane and
+nodes communication.
+
+The suggested approach to Hypershift is to translate the modes as follows:
```

Comment:
what's the rationale for this suggestion? `Public`, `PublicAndPrivate` and `Private` are current options in hcp for aws, they represent cluster endpoints reachability for serving components like kas. Then ingress scope for the data plane is decoupled from the endpoints choice. Also fwiw most operator will run next to the kas being able to access it through the service network, unless they need to run operands in the data plane that need to talk back to the kas.
It'd be a good exercise to pick a few examples and see how they would look like in different form factors

Reply:
TBH this is the most fragile part of this EP, because it's where I have the least understanding (hcp).
From my understanding of previous discussions with @csrwng, configuring an internal HCP cluster does not work the same way it does on OCP.
With HCP, there are three options, while with OCP there are two.
The way I see it these are our options: 
1. translate the choices so that Hypershift can set the Internal/External value on the Infrastructure object based on Public/PublicAndPrivate/Private;
2. change nothing and document the impact;

Option 2 works for me as it simplifies my work in this EP and my use case is currently specific to Azure. 
As long as the impact of exposing cluster publishing status is not restricted by our choice I'm fine with it. Notably, the openshift-apiserver as well as the ingress operator are currently configurable as internal/external at cluster bootstrap time depending on the user choice to publish the cluster or not.

TL;DR: the ultimate idea of this EP is to decentralize public vs private component configuration. If this is not something we want for Hypershift then I would just want to document it here.

Reply:
Thanks for bringing this to light. I'm not aware of the specific use-cases for these different combinations, so I don't know how each of them should affect the registry - I'll need to discuss this with PM.
cc @dmesser

Reply:
> Can we try to detail how this new field would impact the registry operator behaviour (following with your example) for these scenarios.

The registry operator would bootstrap the registry as private when the apiserver is also private. We consider its assets (backing storage) to be private. The registry endpoint would follow whatever ingress configuration the cluster has.

> Can we try to detail how the value/absence of this new field would impact the registry operator in the aws hypershift scenarios?

Currently the registry operator will not look at the cluster publishing status for AWS, so it will not be affected. If it would use this value, it should work as it would for the ARO example: when the apiserver is internal then so will be the assets the creates.